### ISSUE TITLE
[TF 2.0] timestep-wise sample weights not working with tf.data.Dataset and tf.keras.Model

### ISSUE THREAD

**Issue by jrabary** [tensorflow/tensorflow#26803]
**System information**
- TensorFlow version (use command below): tf-nightly-2.0-preview

**Describe the current behavior**

With some model such as object detection one need to weight losses per locations. In this case, the output tensor can have the following shape `[N, T, ...]` where N the batch size and T the locations. 
According to the documentation `tf.keras.Model` can handle timestep-wise sample weight by setting ` sample_weight_mode="temporal"`. 
But this seems not to be the case when we use `tf.Dataset` as inputs.

**Code to reproduce the issue**

```
import tensorflow as tf


class MyModel(tf.keras.Model):

    def __init__(self):
        super(MyModel, self).__init__(name='')

        self.l1 = tf.keras.layers.Conv2D(10, 3)
        self.o1 = tf.keras.layers.Conv2D(2, 1)
        self.o2 = tf.keras.layers.Conv2D(3, 1, name='o2')

        # print(self._output_names)

    def call(self, inputs):
        batch_size = tf.shape(inputs)[0]
        y = self.l1(inputs)
        y1 = self.o1(y)
        y1 = tf.reshape(y1, [batch_size, -1, 2])
        y2 = self.o2(y)
        y2 = tf.reshape(y2, [batch_size, -1, 3])

        return y1, y2


class Loss1(tf.keras.losses.Loss):

    def call(self, targets, predictions):
        losses = tf.math.abs(predictions - targets)
        return tf.reduce_sum(losses, axis=2)


x = tf.random.uniform([10, 16, 16, 3])
y1 = tf.random.uniform([10, 14*14, 2])
y2 = tf.random.uniform([10, 14*14, 3])
w1 = tf.random.uniform([10, 14*14, 1])
w2 = tf.random.uniform([10, 14*14, 1])


data = tf.data.Dataset.from_tensor_slices((x, (y1, y2), {'output_1': w1, 'output_2': w2})).batch(5).repeat()


model = MyModel()
optimizer = tf.keras.optimizers.SGD(0.001)

model.compile(optimizer=optimizer,
              loss=[Loss1(), Loss1()],
              loss_weights=[1., 1.],
              sample_weight_mode="temporal",
              run_eagerly=True)


model.fit(data, epochs=5, steps_per_epoch=1)
```

**Other info / logs**

```
ValueError: Found a sample_weight array with shape (5, 196, 1). In order to use timestep-wise sample weights, you should specify sample_weight_mode="temporal" in compile(). If you just mean to use sample-wise weights, make sure your sample_weight array is 1D.
```

**Comment by mrry** [comment]
From the error message, I think the issue is in Keras. Reassigning to @fchollet for triage.

**Comment by jrabary** [comment]
@fchollet  Is there any advance in this issue. I notice that if we define the model as follow
```
inputs = keras.Input(shape=(16, 16, 3))
y = keras.layers.Conv2D(10, 3)(inputs)
y1 = keras.layers.Conv2D(2, 1)(y)
y1 = keras.layers.Reshape((-1, 2), name='o1')(y1)
y2 = keras.layers.Conv2D(3, 1)(y)
y2 = keras.layers.Reshape((-1, 3), name='o2')(y2)

model = keras.Model(inputs=inputs, outputs=[y1, y2])
```

The timestep-wise sample weight from the dataset works.

It seems that the problem occurs when we subclass `Model` directly. Furthermore the outputs names can't be customised in this case

**Comment by sseveran** [comment]
Any updates on this?

**Comment by caixxiong** [comment]
> @fchollet Is there any advance in this issue. I notice that if we define the model as follow
> 
> ```
> inputs = keras.Input(shape=(16, 16, 3))
> y = keras.layers.Conv2D(10, 3)(inputs)
> y1 = keras.layers.Conv2D(2, 1)(y)
> y1 = keras.layers.Reshape((-1, 2), name='o1')(y1)
> y2 = keras.layers.Conv2D(3, 1)(y)
> y2 = keras.layers.Reshape((-1, 3), name='o2')(y2)
> 
> model = keras.Model(inputs=inputs, outputs=[y1, y2])
> ```
> 
> The timestep-wise sample weight from the dataset works.
> 
> It seems that the problem occurs when we subclass `Model` directly. Furthermore the outputs names can't be customised in this case

I have a similar issue to this. My tf version is 1.14, and I let the `sample_weight=np.zeros(len(X))` as the third element of tf.data . There is no Error happen, but the final loss is not 0, which means that the `sample_weight` argument did not work at all.

**Comment by ymodak** [comment]
This is fixed with TF 2.4.1 see [gist](https://colab.research.google.com/gist/ymodak/94818fff19847334fbcfe594714d7680/untitled17.ipynb) for your reference. Thanks!
